File: src/oop/Stride.java
package oop;

import annotation.Required;
import annotation.HotCode;

/**
 * Off-heap stride registry and byte size lookup utility.
 */
@HotCode
public final class Stride {

    @Required
    public static final int CLASS_ID = TypeRegister.ID_STRIDE;

    private Stride() {}

    public static int classId() {
        return CLASS_ID;
    }

    // get byte stride for class ID
    @HotCode
    @SuppressWarnings("all") // idc if its redundant. the compiler shall treeshake regardless
    public static int get(int generic) {
        int classId = generic & TypeRegister.MASK_CLASS;
        int customStride = Struct.stride(classId);

        if (customStride != 0)
            return customStride;

        return switch (classId) {

            // primtives
            case TypeRegister.ID_BYTE          -> 1;
            case TypeRegister.ID_SHORT         -> 2;
            case TypeRegister.ID_INT,       
                 TypeRegister.ID_FLOAT         -> 4;
            case TypeRegister.ID_LONG,      
                 TypeRegister.ID_DOUBLE,    
                 TypeRegister.ID_STRING,     
                 TypeRegister.ID_INT_FLOAT,
                 TypeRegister.ID_ARGUMENTS     -> 8;
            case TypeRegister.ID_INT_DOUBLE,
                 TypeRegister.ID_LONG_FLOAT,
                 TypeRegister.ID_LONG_DOUBLE   -> 16;

            // variable class, and other outliers
            case TypeRegister.ID_VARIABLE      -> 40;
            case TypeRegister.ID_CLOCK         -> 48;
            case TypeRegister.ID_DATETIME,  
                 TypeRegister.ID_NANOTIME      -> 40;
            case TypeRegister.ID_RANDOM        -> 16;

            // these are just basially wrappers of ones obejct/struct
            case TypeRegister.ID_LIST,      
                 TypeRegister.ID_MAP,       
                 TypeRegister.ID_SET,       
                 TypeRegister.ID_STACK,
                 TypeRegister.ID_DEQUE,
                 TypeRegister.ID_SLAB_ALLOCATOR,
                 TypeRegister.ID_STRING_ENGINE,
                 TypeRegister.ID_SPIN_LOCK,
                 TypeRegister.ID_RING_BUFFER,
                 TypeRegister.ID_MEMORY_MAP_MANAGER,
                 TypeRegister.ID_TRIE,
                 TypeRegister.ID_SEARCH_VARIABLE,
                 TypeRegister.ID_INDEX_RANDOM,
                 TypeRegister.ID_CALENDAR,
                 TypeRegister.ID_GRID_ARRAY,
                 TypeRegister.ID_OCTREE,
                 TypeRegister.ID_CUBE_ARRAY,
                 TypeRegister.ID_SPHERE_ARRAY,
                 TypeRegister.ID_CIRCULAR_ARRAY -> 8; // pointer sizes
            default -> 8;
        };
    }
}

File: src/oop/stride.c
#include "oop/stride.h"

#include "oop/struct.h"
#include "oop/type.h"
#include "annotation/overview.h"

;;OVERVIEW
/**
 * ============================================================================
 * MODULE: Stride (oop/stride.c)
 * LEVEL: L1 — File Metadata (class byte-width metadata)
 * ============================================================================
 * the Stride utility, ported from oop/Stride.java.
 *
 * STRUCT FIELDS: none — procedural (operates on Class/Struct stride registry (no instance state))
 *
 * FUNCTION REGISTRY:
 * ----------------------------------------------------------------------------
 * Getters:
 *   - Stride_get(class_id)
 * ============================================================================
 */


// stride.c — Stride port (Legacy: oop/Stride.java). Byte width of a class.

size_t Stride_get(uint32_t class_id) {
    uint64_t id = (uint64_t)class_id & MASK_CLASS;

    // Runtime-defined custom structs first: the Struct registry is the source
    // of truth for their stride.
    if (id >= ID_CUSTOM_STRUCT) {
        size_t custom = Struct_stride(id);
        if (custom != 0)
            return custom;
    }

    switch (id) {
        case ID_BYTE:       return 1;
        case ID_SHORT:      return 2;
        case ID_INT:
        case ID_FLOAT:
        case ID_BIT8:
        case ID_BIT16:
        case ID_BIT32:      return 4;
        case ID_LONG:
        case ID_DOUBLE:
        case ID_FIXED64:
        case ID_BIT64:
        case ID_STRING:
        case ID_INT_FLOAT:
        case ID_ARGUMENTS:  return 8;
        case ID_INT_DOUBLE:
        case ID_LONG_FLOAT:
        case ID_LONG_DOUBLE: return 16;

        // Variable and the time outliers.
        case ID_VARIABLE:   return 40;
        // Math / geometry blocks.
        case ID_VEC2:       return 8;
        case ID_VEC3:       return 12;
        case ID_VEC4:       return 16;
        case ID_MAT3:       return 36;
        case ID_MAT4:       return 64;
        case ID_QUATERNION: return 16;
        case ID_ENTITY:     return 8;

        // Wrappers of a single pointer/struct.
        case ID_RANDOM:     return 16;
        case ID_LIST:
        case ID_MAP:
        case ID_SET:
        case ID_STACK:
        case ID_DEQUE:
        case ID_QUEUE:
        case ID_MIN_HEAP:
        case ID_SPARSE_SET:
        case ID_SPINLOCK:
        case ID_RING_BUFFER:
        case ID_INDEX_RANDOM: return 8;
        default:            return 8;
    }
}
